---
title: "Analysis of Prostate Cancer Exome data reveals novel biomarkers"
author: "Yogesh Maithania"
output: 
  html_document:
    toc: true
    theme: cerulean
    df_print: paged
    number_sections: true
---

```{r setup, include = FALSE}

knitr::opts_chunk$set(message = FALSE, warning = FALSE)

library(dplyr)
library(ggplot2)
```

**Question 1**

Download the dataset linked to above and load the Mutation Annotation Format (MAF) files found in the data/mafs/ folder. Each of these 50 files contains the genomic mutations observed in a different patient’s tumor, obtained by biopsy and sequenced with whole-exome sequencing. Each row in a MAF file corresponds to a different mutation.

```{r read_in_and_organize_data}
## load patient Information
patient_file <- read.table("sample-information.tsv", header = TRUE)

# store Response variable
responder_ids <- patient_file$Patient_ID[patient_file$Response == "Responder"]
non_responder_ids <- patient_file$Patient_ID[patient_file$Response == "Non-Responder"]

## processing each of the patient MAF files separately to focus on patient-specific characteristics
# list all MAF files in the directory and sort them numerically
maf_files <- list.files(path = "mafs/", full.names = TRUE)
maf_files <- maf_files[order(as.numeric(gsub("\\D", "", maf_files)))]

# extract patient IDs for list index
patient_names <- paste0("Patient-", 0:(length(maf_files) - 1))

# read 
maf_data_list <- lapply(maf_files, read.table, header = TRUE, comment.char = "#", 
                        sep = "\t", stringsAsFactors = FALSE)
names(maf_data_list) <- patient_names

# check
# str(maf_data_list[['Patient-27']])
```


**Question 2**

Subset for mutations that are not of the Variant Classification “Silent”. For the purposes of this analysis, we will restrict ourselves to substitutions which result in changes to the produced protein (“nonsynonymous mutations”).

```{r subset_variant_class}
# drop rows that includes "Silent" variant class in all MAF files
maf_data_list_subset <- lapply(maf_data_list, function(df) subset(df, Variant_Classification != "Silent"))

# verify results
unique(maf_data_list_subset[['Patient-27']]$Variant_Classification)
```


**Question 3**

Find the 15 most common mutations. Gene names are included in the column Hugo_Symbol and protein changes are stored in the column Protein_Change.

```{r 15_common_mutations}
# combine all patient data based on mutational gene and protein change
combined_maf_data <- maf_data_list_subset %>%
  lapply(function(df) select(df, Hugo_Symbol, Protein_Change)) %>%
  bind_rows(.id = "Patient_ID")
# head(combined_maf_data)

# identify 15 common mutations based on occurrences across all patient data
top15_mutations <- combined_maf_data %>%
  group_by(Hugo_Symbol, Protein_Change) %>%
  summarise(Frequency = n(), Patients = paste(unique(Patient_ID), collapse = ", "), .groups = "drop") %>%
  arrange(desc(Frequency)) %>%
  head(15)

# check results
top15_mutations
```
The data reveals an interesting pattern in the frequency of mutations across various genes. Notably, ***ERBB4*** stands out with two occurrences, showcasing the gene's susceptibility to mutations, each involving a distinct protein change. This suggests a potential functional diversity or selective pressure on ***ERBB4*** Moreover, the uniform range of frequency among other genes implies a balanced distribution of mutations. However, the presence of one exceptionally higher frequency, as seen in a specific gene or mutation, could indicate its significance or possible association with a particular condition.


**Question 4**

Perform a statistical test to explore if any mutated genes are enriched in patients who either responded or not. Response labels for individual patients are found in the file data/sample-information.tsv.

Here, we perform the Fisher's exact test to determine if certain gene mutations show significant enrichment in patients who responded or did not respond to treatment. Using contingency tables for each gene mutation, the test evaluates if the observed distribution of mutations among responders and non-responders is statistically significant, considering the categorical nature of the data and small sample size.

```{r statistical_testing}
# initialize our results data frame
test_results <- data.frame()
# separate Patient IDs for each gene
top15_mutations$Patients <- strsplit(as.character(top15_mutations$Patients), ", ")

# iterate through each row in top15_mutations, calculate Fisher's exact test for
# responders and non-responders, and store results
for (i in 1:nrow(top15_mutations)) {
  gene <- top15_mutations$Hugo_Symbol[i]
  mutation <- top15_mutations$Protein_Change[i]
  frequency <- top15_mutations$Frequency[i]
  patients <- unlist(top15_mutations$Patients[i])
  
  # identify patient IDs for each gene
  responders <- intersect(responder_ids, patients)
  non_responders <- intersect(non_responder_ids, patients)
  
  # construct the table
  cont_table <- matrix(c(length(responders), length(setdiff(patients, responders)),
                         length(non_responders), length(setdiff(patients, non_responders))), 
                       nrow = 2, byrow = TRUE)
  
  # perform Fisher's Exact test
  fisher_result <- fisher.test(cont_table)
  
  # store results
  test_results <- rbind(test_results, c(gene, mutation, frequency, 
                              round(fisher_result$p.value, 4)))
}

# rename the columns
colnames(test_results) <- c("Hugo_Symbol", "Protein_Change", "Frequency", "PValue")

test_results <- test_results %>%
  mutate(
    Frequency = as.numeric(Frequency),
    PValue = as.numeric(PValue)
  )

# check results
test_results
```
The statistical test results provide additional depth to the interpretation of the mutation frequency data. Interestingly, despite ERBB4 having two occurrences, the p-values associated with the mutations (p.S1289A and p.Q707E) are relatively high (0.7064 and 1.0000, respectively). This may suggest that these mutations are not statistically significant, questioning their potential impact or association with certain conditions. In contrast, ***KRTAP4-11, MAP2K1, and TYRO3*** mutations exhibit significantly low p-values (0.0286), indicating a higher likelihood that these mutations are not mere random events. They have identical p-values ***(0.0286)*** that might suggest a consistent pattern in treatment response distribution among patients with these mutations. This finding might imply a potential shared impact of these genetic alterations on the response to treatment in the studied population.


**Question 5**

Create a scatter plot of genes with the number of mutated patients on the x-axis and your results from question 4 on the y-axis. Can the figure in any way to improve readability? If so, recreate the plot using your suggestion(s).

```{r scatterplot, fig.height = 5, fig.width = 7}
# representing the relationship between the number of mutated patients and the statistical significance of gene mutations 
# setting color codes to highlight significant genes
plot_data <- test_results
plot_data$Color <- ifelse(plot_data$PValue < 0.05, "green", "red")

# scatter plot
ggplot(plot_data, aes(x = Frequency, y = PValue, color = Color)) +
  geom_point(alpha = 0.7, size = 3) +
  geom_text(aes(label = Hugo_Symbol), position = position_jitter(width = 0.2, height = 0.2), size = 3) +
  scale_color_identity() +
  labs(title = "Understanding Mutation Occurrence Across 50 Patients (using 15 mutations)",
       x = "Number of Mutated Patients",
       y = "P-Values",
       title.theme = element_text(hjust = 0.5)) + 
  theme_minimal()
```


**Question 6**

(a) How many samples are wild-type versus mutant with respect to the most significantly enriched gene from Question 4? 
(b) Plot the number of non-synonymous mutations per megabase in the mutant vs. wild-type samples. 
(c) Is there a significant difference in the number of mutations between the two groups? 
Information on the number of non-synonymous mutations per megabase for each patient can be found in the file data/sample-information.tsv.

```{r mutant_vs_wildtype_a}
##(a)##
# extract statistically significant genes/mutations that are enriched
sig_enriched_genes <- test_results[test_results$PValue < 0.05, "Hugo_Symbol"]
sig_enriched_genes

# get corresponding patient IDs i.e., mutant samples
get_patient_ids <- lapply(sig_enriched_genes, function(gene) {
  patients <- top15_mutations$Patients[top15_mutations$Hugo_Symbol == gene]
})
mutant_samples <- unique(unlist(get_patient_ids))
mutant_samples
```
From our analysis on 15 gene mutations, out of 50 samples, we may have 44 wild-type samples.


```{r mutant_vs_wildtype_b,  fig.height = 4, fig.width = 6}
##(b)##
# Filter patient file data for mutant and wild-type samples
mutant_data <- patient_file %>% filter(Patient_ID %in% mutant_samples)
wild_type_data <- patient_file %>% filter(!Patient_ID %in% mutant_samples)

# box plot
ggplot() +
  geom_boxplot(data = mutant_data, aes(x = "Mutant", y = Nonsynonymous_mutations_per_Mb), fill = "lightblue") +
  geom_boxplot(data = wild_type_data, aes(x = "Wild-Type", y = Nonsynonymous_mutations_per_Mb), fill = "lightgreen") +
  labs(title = "Nonsynonymous Mutations per Megabase",
       x = "Samples",
       y = "Nonsynonymous Mutations per Mb") +
  theme_minimal()
```

The box plot analysis reveals distinct patterns in Nonsynonymous Mutations per Megabase  between mutant and wild-type samples. In the mutant group, a smaller box with a median close to the lower quartile indicates a concentration of samples with a lower mutation rate, possibly suggesting a specific mutation profile. Conversely, in the wild-type group, a wider box with a pronounced gap between the median and top quartile suggests a broader range of mutation rates, with some samples exhibiting higher mutational burdens. The presence of outliers in both groups signifies individual samples with exceptionally high mutation rates. Evidently, their is no substantial overlap between the inter-quartile range and medians are different, it may imply that the mutations observed in the mutant group can be significantly different from those in the wild-type group.

```{r mutant_vs_wildtype_c}
## validation of the box plot
# testing significance between the two groups 
mutant_vs_wt_test <- t.test(mutant_data$Nonsynonymous_mutations_per_Mb, 
                       wild_type_data$Nonsynonymous_mutations_per_Mb)
mutant_vs_wt_test
```
Yes, the t-test results demonstrate a highly significant difference in the number of mutations between the mutant and wild-type groups (p-value << 0.05). The lower mean mutation rate in the mutant group, supported by the negative t-value and confidence interval, solidifies this finding.


**Question 7**

Write any conclusions that you have made based on your analysis. How might this analysis be improved or expanded upon? Please include all requested figures in your report.

Based on the analysis, three genes stand out as significantly enriched: ERBB4, KRTAP4-11, and MAP2K1. These genes exhibit distinct mutation patterns, especially in the context of treatment response. ERBB4, despite having two occurrences, does not show statistical significance in the t-test, suggesting that the mutations observed may not be strongly associated with treatment response. However, KRTAP4-11 and MAP2K1 mutations significantly differ in mutation rate between mutant and wild-type samples, indicating a potential association with treatment response.
Given that our analysis focused solely on the top 15 common mutations, it is crucial to consider expanding the scope to include all enriched gene mutations for more comprehensive statistical testing. This approach would provide a holistic view of the mutational landscape, potentially revealing additional genes associated with treatment response.
Moving forward, conducting pathway or molecular analyses, such as Gene Set Enrichment Analysis (GSEA), could emphasize more on the biological pathways influenced by these mutations.

