---
title: "Bulk RNA-seq analysis of pediatric localized scleroderm"
author: "YM"
date: "2024-07-27"
output:
  rmdformats::readthedown:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    highlight: tango  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE, warning=FALSE}
## func to install and load pkgs from parent repo
install_and_load <- function(pkgs) {
  if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  
  new_pkg <- pkgs[!(pkgs %in% installed.packages()[,"Package"])]
  if(length(new_pkg)) BiocManager::install(new_pkg, update = FALSE, ask = FALSE)
  
  lapply(pkgs, library, character.only = TRUE)
}

packages <- c("rmdformats", "here", "dplyr", "edgeR", "limma", "ggplot2", "ggrepel", "ComplexHeatmap", "Rtsne", 
              "fgsea", "msigdbr", "clusterProfiler", "writexl", "org.Hs.eg.db", "enrichplot")

## call the func
invisible(install_and_load(packages))
```

## Data exploration
```{r}
## read and explore the data
raw_counts <- read.csv(gzfile(here("Data/GSE166861_FPKM_normalized_data.csv.gz")))

## convert to matrix
count_matrix <- raw_counts[-1] %>% as.matrix
rownames(count_matrix) <- raw_counts$gene_id

## replace NA with column means
col_means <- colMeans(count_matrix, na.rm = TRUE)
counts_mat_filled <- t(apply(count_matrix, 1, function(x) {
  ifelse(is.na(x), col_means, x)
}))

## glance the raw data
count_matrix <- counts_mat_filled[rowSums(!is.na(count_matrix)) > ncol(count_matrix)/2, ]
head(count_matrix)
```

```{r}
## define metadata
coldata <- data.frame(Samples = colnames(count_matrix),
                      Condition = ifelse(grepl("^HSK", colnames(count_matrix)), "Healthy", "LS"))

coldata
```

## Library Distribution
```{r, library distribution}
## convert counts to DGE list
dge <- DGEList(counts=count_matrix)

## add group information
dge$samples$group <- coldata$Condition
sample_names <- coldata$Samples

## filter low-expressed genes
## keeping genes with at least 1 count-per-million (CPM) in at least 2 samples
keep <- filterByExpr(dge, group = dge$samples$group)
dge <- dge[keep, , keep.lib.sizes = FALSE]

## normalize counts using TMM normalization
dge <- calcNormFactors(dge)

## enhanced plots of library sizes before and after normalization
par(mar = c(6, 5, 4, 2)) 

## barplot for raw library sizes
raw_lib_sizes <- dge$samples$lib.size 
mean_raw_lib_size <- mean(raw_lib_sizes)

## create a plotting df
data1 <- data.frame(Sample = sample_names, LibrarySize = raw_lib_sizes)

## plot
rls_plot <- ggplot(data1, aes(x = Sample, y = LibrarySize)) +
  geom_bar(stat = "identity", fill = "skyblue") + 
  geom_hline(yintercept = mean_raw_lib_size, color = "red", linetype = "dashed", linewidth = 0.5) + 
  labs(
    title = "Library distribution",
    x = "Samples", 
    y = "Library Size"
  ) +
  theme_minimal() +
  theme(
    plot.background = element_rect(fill = "white"),   
    axis.text.x = element_text(angle = 45, hjust = 1),
    plot.title = element_text(hjust = 0.5, face = "bold") 
  ) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1))) +
  annotate("text", x = 1, y = mean_raw_lib_size + 5000, label = "", color = "red", size = 2)

## save
ggsave("raw_libsize_LS_vs_HC.png", rls_plot, width = 12, height = 10, units = "in")
## view
print(rls_plot)

## barplot for normalized library sizes
norm_lib_sizes <- dge$samples$lib.size * dge$samples$norm.factors
mean_norm_lib_size <- mean(norm_lib_sizes)

## create a plotting df
data2 <- data.frame(Sample = sample_names, LibrarySize = norm_lib_sizes)

## plot
nls_plot <- ggplot(data2, aes(x = Sample, y = LibrarySize)) +
  geom_bar(stat = "identity", fill = "lightgreen") + 
  geom_hline(yintercept = mean_norm_lib_size, color = "red", linetype = "dashed", linewidth = 0.5) + 
  labs(
    title = "Library distribution",
    x = "Samples", 
    y = "Library Size"
  ) +
  theme_minimal() +
  theme(
    plot.background = element_rect(fill = "white"),
    axis.text.x = element_text(angle = 45, hjust = 1),
    plot.title = element_text(hjust = 0.5, face = "bold") 
  ) +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1))) +
  annotate("text", x = 1, y = mean_norm_lib_size + 5000, label = "", color = "red", size = 2)

## save
ggsave("norm_libsize_LS_vs_HC.png", nls_plot, width = 12, height = 10, units = "in")
## view
print(nls_plot)

```

## Principal Component Analysis
```{r PCA}
## Calculate covariance matrix  
pca_res <- prcomp(t(dge$counts), center = TRUE, scale. = TRUE)
explained_variance <- summary(pca_res)$importance[2, ]

## pca dataframe having calculated PCs (scores, eigenvectors, eigenvalues)
pca_df <- as.data.frame(pca_res$x)
pca_df$Sample <- coldata$Condition 

## plot 
pca_plot <- ggplot(pca_df, aes(x = PC1, y = PC2, label = "")) +
  geom_point(size = 3, aes(color = Sample)) +
  geom_text(hjust = 1.5, vjust = 1.5) +
  labs(
    title = "PCA Plot",
    x = paste0("PC1 (", round(explained_variance[1] * 100, 2), "%)"),
    y = paste0("PC2 (", round(explained_variance[2] * 100, 2), "%)")
  ) +
  theme_minimal() + 
  theme(plot.background = element_rect(fill = "white")
  )

## save
ggsave("pca_plot_LS_vs_HC.png", pca_plot, width = 12, height = 10, units = "in")
## view
print(pca_plot)
```

## t-Distributed Stochastic Neighbor Embedding 
```{r tSNE}
set.seed(27)
tsne_res <- Rtsne(t(dge$counts), perplexity = 5, dims = 2, verbose = TRUE, max_iter = 1000)

## tSNE dataframe 
tsne_df <- as.data.frame(tsne_res$Y)
colnames(tsne_df) <- c("tSNE1", "tSNE2")
tsne_df$Sample <- coldata$Condition

## plot
tsne_plot <- ggplot(tsne_df, aes(x = tSNE1, y = tSNE2, label = "")) +
  geom_point(size = 3, aes(color = Sample)) + 
  geom_text(hjust = 1.5, vjust = 1.5) +
  labs(
    title = "t-SNE Plot",
    x = "tSNE1",
    y = "tSNE2"
  ) +
  theme_minimal() + 
  theme(plot.background = element_rect(fill = "white")
  )

## save
ggsave("tsne_LS_vs_HC.png", tsne_plot, width = 12, height = 10, units = "in")
## view
print(tsne_plot)
```

## Differential Expression Analysis
```{r, DGE}
## design matrix for group comparisons
condition <- factor(c(rep("Healthy", 4), rep("LS", 14)))
design_mat <- model.matrix(~condition)

## convert to log2 scale and filter low expression genes
log2_fpkm <- log2(counts_mat_filled + 1)
keep <- rowSums(log2_fpkm > 0) >= 3  # Keep genes expressed in at least 3 samples
filtered_log2_fpkm <- log2_fpkm[keep,]

## fit linear model
fit <- lmFit(filtered_log2_fpkm, design_mat)

## apply eBayes
fit <- eBayes(fit)

## save differentially expressed genes
results <- topTable(fit, number = Inf, adjust.method = "BH", sort.by = "P")

## save statistically significants DEGs
sig_genes <- results[results$adj.P.Val < 0.05,]

## get up, and down regulated 
deg_up <- subset(results, logFC > 0 & adj.P.Val < 0.05)
deg_down <- subset(results, logFC < 0 & adj.P.Val < 0.05)

## save them
output_file <- here("HC_vs_LS_deg.xlsx")
write_xlsx(
  list(
    "All" = sig_genes,
    "Up" = deg_up,
    "Down" = deg_down
  ),
  path = output_file
)

## create a df based on cutoffs
deg_data <- results %>%
  mutate(
    regulated = case_when(
      adj.P.Val < 0.05 & logFC > 0 ~ "up",
      adj.P.Val < 0.05 & logFC <= 0 ~ "down",
      TRUE ~ "not significant"
    )
  )

# comparison between DEGs
table(deg_data$regulated)
```

## Volcano plot
```{r, volcano}
top_genes <- deg_data %>%
  filter(regulated != "not significant") %>%
  group_by(regulated) %>%
  arrange(adj.P.Val, desc(abs(logFC))) %>%
  slice_head(n = 10) %>%
  ungroup()

volcano_plot <- ggplot(deg_data, aes(x = logFC, y = -log10(adj.P.Val))) +
  geom_point(aes(color = regulated), alpha = 0.7, size = 2) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black") +
  geom_point(data = top_genes, size = 3, shape = 21, fill = "white") + ## highlight top genes
  ## add labels for top genes
  geom_text_repel(
    data = top_genes,
    aes(label = ID),
    size = 3,
    box.padding = 0.5,
    point.padding = 0.5,
    force = 8,
    max.overlaps = Inf
  ) +
  ## add colors and them
  scale_color_manual(values = c("up" = "red", "down" = "blue", "not significant" = "grey")) +
  labs(
    title = "Volcano Plot of DEGs",
    subtitle = paste0("Top 10 up- and down-regulated genes highlighted"),
    x = "Log2 Fold Change",
    y = "-Log10 Adjusted P-value",
    color = "Regulation"
  ) +
  theme_bw(base_size = 12) +
  theme(
    legend.position = "right",
    plot.title = element_text(hjust = 0.5, face = "bold"),
    plot.subtitle = element_text(hjust = 0.5),
    axis.title = element_text(face = "bold")
  )

ggsave("volcanoplot_LS_vs_HC.png", volcano_plot, width = 12, height = 10, units = "in")

# view
volcano_plot
```

## Heatmap for significant genes
```{r, heatmap, warning=FALSE}
## condition df
condition <- factor(c(rep("Healthy", 4), rep("LS", 14)))

## annotation df for mapping
annotation_col <- data.frame(
  condition = condition,
  row.names = colnames(filtered_log2_fpkm)
)

## subset the expression data for DEGs
deg_expression <- filtered_log2_fpkm[rownames(filtered_log2_fpkm) %in% sig_genes$ID, ]

## saving plot

png("heatmap_LS_vs_HC.png", width = 2000, height = 2400, res = 300)
## create the heatmap
hp_sigs <- pheatmap(
  deg_expression,
  annotation_col = annotation_col,
  show_rownames = FALSE,
  show_colnames = FALSE,
  main = "Heatmap of DEGs",
  color = colorRampPalette(c("navy", "white", "firebrick3"))(100),
  fontsize = 10,
  scale = "row",
  annotation_colors = list(condition = c("LS" = "darkgreen", "Healthy" = "orange")),
  annotation_legend = TRUE,
  treeheight_col = 0,
  treeheight_row = 50
)
print(hp_sigs)
dev.off()

draw(hp_sigs)
```

## GO Enrichment Analysis
```{r, GO}
## GO analysis for up and down regulated genes
## define func
ego <- function(genes, title) {
  ## perform GO enrichment
  go_result <- enrichGO(
    gene = genes,
    OrgDb = org.Hs.eg.db,
    keyType = "SYMBOL",
    ont = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff = 0.05,
    qvalueCutoff = 0.2,
    readable = TRUE
  )
  
  ## create barplot
  bar_go <- barplot(go_result,
                    showCategory = 20,
                    title = title,
                    font.size = 10,
                    x = "GeneRatio",
                    color = "pvalue")
  
  return(list(result = go_result, plot = bar_go))
}

## sort and get DE genes, call the func
deg_sorted <- deg_up[order(deg_up$logFC, decreasing = TRUE), ]
gene_list_up <- deg_sorted$ID
up_go <- ego(gene_list_up, "Up regulated GO enrichment")

## plot and save
print(up_go$plot)
ggsave("GO_bar_up.png", up_go$plot, width = 12, height = 8, dpi = 300)

## sort and get DE genes, call the func
deg_sorted2 <- deg_down[order(deg_down$logFC, decreasing = TRUE), ]
gene_list_down <- deg_sorted2$ID
down_go <- ego(gene_list_down, "Down regulated GO enrichment")

## plot and save
print(down_go$plot)
ggsave("GO_bar_down.png", down_go$plot, width = 12, height = 8, dpi = 300)
```

## Gene set enrichment Analysis
```{r, GSEA, warning=FALSE}
## get ranked genes
gene_list <- sig_genes$logFC
names(gene_list) <- sig_genes$ID
ranked_genes <- sort(gene_list, decreasing = TRUE)

## load gene sets as our background
m_df <- msigdbr(species = "Homo sapiens", category = "H")
gene_sets <- split(m_df$gene_symbol, m_df$gs_name)

## run GSEA
gsea_results <- fgsea(pathways = gene_sets, 
                      stats = ranked_genes,
                      minSize = 15,
                      maxSize = 500,
                      nperm = 10000)

## sort results by NES 
gsea_results <- gsea_results %>% arrange(desc(NES))

## barplot plot for pathways
gsea_bar <- ggplot(gsea_results, aes(reorder(pathway, NES), NES)) + 
  geom_col(aes(fill = padj < 0.05), show.legend = FALSE) + 
  coord_flip() +  
  labs(x = "Pathway", 
       y = "NES", 
       title = "Top pathways") +  
  theme_minimal(base_size = 12) + 
  theme(plot.background = element_rect(fill = "white"),
        axis.text.y = element_text(size = 10),  
        plot.title = element_text(hjust = 0.5, face = "bold")
        )

## save
ggsave("GSEA_barplot_LS_vs_HC.png", gsea_bar, width = 12, height = 8, dpi = 300)

## view
print(gsea_bar)

## enrichment plot
top_pathway <- gsea_results$pathway[[1]]
gsea_enrich <- plotEnrichment(gene_sets[[top_pathway]], ranked_genes) + 
  labs(title = top_pathway)

## save GSEA plots
ggsave("GSEA_enrich_LS_vs_HC.png", gsea_enrich, width = 12, height = 8, dpi = 300)

## view
print(gsea_enrich)
```